.eslintrc.js

```diff
@@ -1,4 +1,7 @@
-module.exports = {
+// @ts-check
+const { typedConfig } = require('@typescript-eslint/typed-config');
+
+module.exports = typedConfig({
```

Comment:
q: do we have to wrap it like this?

in typescript you can annotate js code with comments eg.

```js
// @ts-check

/** @type {Record<string, number>} */
const config = {
  foo: 'test'
};

module.exports = config
```

it means that something like this should work (not tested)

```js
/// <reference path="@typescript-eslint/typed-config" />
// @ts-check

/** @type {ESLintConfig.ConfigWithOverrides} */
const config = {
  foo: 'test'
};

module.exports = config
```

there is advantage in this case of using jsdoc, as `@typescript-eslint/typed-config` will not be required at runtime for end users

Reply:
this could go to docs for plugin creators

Reply:
Yeah there's just a tradeoff between the two. I'm not worried about needing the package at runtime for a few reasons:
- The function will just be a no-op, so it's more or less zero impact.
- The function abstracts away the exact type names, so we can change the type names later without worrying as much.
- Finally, a niceness which we may or may not leverage: having a function means that in future, we can consider adding in some nice transformations to the config without having people migrate to a function.

My gut says that most people don't really understand reference comments or how to use imported types in JSDoc comments. Only because most people wouldn't have used it before.
Though at the same time, most people will just copy paste code from our documentation into their config, so this isn't a huge deal breaker.

Either approach will work I think.

Reply:
maybe we should just support both, function for normal users and jsdoc for plugin creators (that they don't have to include this package in dependencies)

Reply:
@armano2 while you can use JSDoc, I think there is an advantage to having a function like this too, including:

* You don't need to use a variable purely for "meta" reasons - by no means the end of the world, but still one step away from the "optimal" setup
* IDEs can have better interop - often JSDoc & TS are handled by different subcomponents; this tends to manifest most in the form of JSDoc being "weaker" than TS i.e WebStorm will bail out or ignore JSDoc more than it would TS (including from type definitions)
* It's something the end consumer doesn't have to maintain: I'm calling  function that takes are argument - my tooling knows the details. End of story. Using a JSDoc via a variable means you've now got this comment that is semi-magically doing something, it's tied to the variable so if you move it around you need to take it with you, and it's more visible which I think is the opposite of what we'd ideally like (which is why I proposed `.eslintrc.ts` in the first place).

I agree that supporting & having both documented should be a focus, but that's the advantage I see in the function :)

Reply:
> Oohh I did not know about that! 😄

i just read something about this, (i'm unsure if syntax that i pointed out is correct)

https://github.com/eslint/eslint/pull/12321/files

https://github.com/evanplaice/eslint/blob/dd1c30e35f05ed332e2abbd3d4d53635efde74b8/tests/lib/cli-engine/config-array-factory.js#L1494-L1502

Reply:
@armano2 haha actually this would make it even harder: that works for each top level configuration property:

```
exports.env = { node: true };
exports.parser = '@typescript-eslint/parser';
exports.parserOptions = {
  project: 'tsconfig.eslint.json',
  sourceType: 'module',
  ecmaVersion: 2019
};
exports.plugins = ['@typescript-eslint/eslint-plugin', 'local'];
exports.extends = ['./index.js', './@typescript-eslint.js'];
exports.overrides = [
  { files: ['*.spec.*'], extends: ['./jest.js'] },
  {
    files: ['*.js'],
    rules: {
      '@typescript-eslint/no-require-imports': 'off',
      '@typescript-eslint/no-var-requires': 'off'
    }
  }
];
exports.rules = {
  '@typescript-eslint/no-namespace': 'off',
  '@typescript-eslint/no-dynamic-delete': 'off',
  'local/prefer-ast-types-enum': 'error',
  'local/prefer-valid-rules': 'error',
  'no-sync': 'off'
};
```

Reply:
> jsdoc for plugin creators (that they don't have to include this package in dependencies)

How would jsdoc work for the plugin creators? How can you do an interface declaration merge without having the package in dependencies?

My thinking was that the `typed-config` package would be a package that end users could install to expose the function.
The interfaces themselves would be defined instead in `experimental-utils`, which plugin authors will already be using if they care about typescript.